### PR TITLE
Update data representation, allow for remote authors and remote likes

### PR DESCRIPTION
**Pull request by debecerra**
### Summary 
- Split Author model into Author (for all authors) and LocalAuthor (for only local authors) where Author is a parent class of LocalAuthor
- Created Like models to represent likes on Posts and Comments
- Relates to #108 but does not completely address it. Still need to update data representation for Posts
- Liking posts as a remote author now works. Still need to make a few more changes to allow for the following/befriending functionality to work for remote authors. 
- Need to make similar changes for receiving remote posts

### Notes
 - `Author` was renamed to `LocalAuthor`, and `Author` is now a parent class of `LocalAuthor` since a `LocalAuthor` "is a" `Author`. I added lots of documentation to these classes to explain their purpose and reason for this design.
 - Remote authors can like a post. I created a test that demonstrates how this works. You can use the request body found in the test to play with it yourself. Essentially, the actor (author) is queried in the Author table by their URL ID. If this Author is LocalAuthor, their ID will be found. If the author is not local, we create a new Author instance to represent this remote user. The Like object is created with a reference to this Author object (which is not a LocalAuthor). 
 - Suppose the same author likes another post. The `get_or_create` lookup will find their URL ID in the database, so the same Author instance is used to create the new Like. 
 - A local author can be queried from both Author and LocalAuthor models. The author will have the same ID in both tables. This makes it easy to go from Author to LocalAuthor (if one exists).

Class Diagram: https://drive.google.com/file/d/1uP1oHBm3zoss30xdh8-sijCBG--XY4JM/view?usp=sharing which is the overall idea of what I am aiming for with this. 

@JawadHossain Can you specifically make sure you review this? I had a lot of conflicts with the changes you pushed in your last PR. Want to make sure I didn't break anything. Thanks

### CODE REVIEW

File: code/api/views.py [hunk 1]
```diff
@@ -331,57 +332,70 @@ def post(self, request, author_id):
...
+
+                # extract data from request body
                 splitObject = data["object"].split("/")
-                object = data["object"].split("/")[-2]
-                id = splitObject[-1]
+                object = data["object"].split("/")[-2]  # Feels a bit dangerous, what if there is a trailing '/'?
+                id = splitObject[-1] # Dangerous
+                liking_author_url = data["author"]["id"]
+
+                # let's try to use url_parser.parse_something(data["object"]) for parsing url
```
Comment (JawadHossain) [inline code/api/views.py:358]: okay, thanks

**Review by JawadHossain — APPROVED** [review]
Awesome work!  
Do we need socialDistribution/models/follow.py ?

File: code/api/views.py (hunk 1 above)
Comment (JawadHossain) [reply code/api/views.py:358]: parsing url now 4deeb34b413e19483cde033eca88eff7fc8f7477. 

Thanks a lot for pulling from my branch! I think you already have resolved all merge conflicts. Can you please also keep the URL parser code from here after I merge?

I think @tylerelston added the reference on line 349. Was this the intended place for it?

**Review by olivier-vadiaval — APPROVED** [review]
Looks very good! Your design is well thought through 💯

**Comment by debecerra** [comment]
> Awesome work! Do we need socialDistribution/models/follow.py ?

Good point! I initially planned to use this for a FriendRequest model but didn't get to it in this PR. I'll just delete that file to keep things clean and add it if/when needed.

File: code/api/views.py (hunk 1 above)
Comment (debecerra) [reply code/api/views.py]: Removed my comment for now. @tylerelston if you could update the reference with some added context and make sure it's on the correct line on your next PR that would be great.

**Pull request merged**